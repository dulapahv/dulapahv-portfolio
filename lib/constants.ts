import type { COBEOptions } from 'cobe';

export const IS_DEV_ENV =
  process.env.VERCEL_ENV === 'development' ||
  process.env.NEXT_PUBLIC_ENV === 'development' ||
  process.env.NODE_ENV === 'development';
export const BASE_URL = IS_DEV_ENV ? 'http://localhost:3000' : 'https://dulapahv.dev';
export const ASSETS_URL = 'https://assets.dulapahv.dev';
export const LINKEDIN_URL = 'https://www.linkedin.com/in/dulapahv';
export const GITHUB_URL = 'https://github.com/dulapahv';

export const NAME = 'Dulapah Vibulsanti';
export const DESCRIPTION =
  'Thai Software Engineer based in Edinburgh, United Kingdom. Passionate about creating accessible and delightful technology for everyone.';

export const NAME_MAX_LENGTH = 256;
export const EMAIL_MAX_LENGTH = 256;
export const MESSAGE_MAX_LENGTH = 2000;

export const EMAIL_REGEX = /^\S+@\S+\.\S+$/;
export const TURNSTILE_SITE_KEY = IS_DEV_ENV
  ? '1x00000000000000000000AA'
  : '0x4AAAAAAACYFWWcTzhCNWz4';

export const EDUCATION_LOCATION: COBEOptions['markers'] = [
  // KMITL, Thailand
  { location: [13.7298889, 100.775652], size: 0.1 },

  // UofG, United Kingdom
  { location: [55.8716946, -4.2913655], size: 0.1 }
];
