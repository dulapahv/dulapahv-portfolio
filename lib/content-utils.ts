import { allBlogs, allProjects, allWorks } from 'content-collections';

export type ContentType = 'project' | 'blog' | 'work';

export type ContentItem =
  | (typeof allProjects)[0]
  | (typeof allBlogs)[0]
  | (typeof allWorks)[0];

export const contentConfig = {
  project: {
    title: 'Projects',
    description: 'Professional and personal projects I have worked on.',
    collection: allProjects,
    label: 'Project',
    pluralLabel: 'Projects',
  },
  blog: {
    title: 'Blog',
    description: 'My thoughts, ideas, and experiences.',
    collection: allBlogs,
    label: 'Article',
    pluralLabel: 'Articles',
  },
  work: {
    title: 'Work',
    description: "Companies and clients I've worked with.",
    collection: allWorks,
    label: 'Case Study',
    pluralLabel: 'Case Studies',
  },
} as const;

export const isValidContentType = (type: string): type is ContentType => {
  return type === 'project' || type === 'blog' || type === 'work';
};

export const getCollection = (type: ContentType) => {
  return contentConfig[type].collection;
};

export const getAllContent = (): Array<ContentItem & { type: ContentType }> => {
  const projects = allProjects.map((item) => ({
    ...item,
    type: 'project' as ContentType,
  }));
  const blogs = allBlogs.map((item) => ({
    ...item,
    type: 'blog' as ContentType,
  }));
  const works = allWorks.map((item) => ({
    ...item,
    type: 'work' as ContentType,
  }));

  return [...projects, ...blogs, ...works];
};

// Helper function to get the relevant date for sorting
const getRelevantDate = (item: ContentItem): Date => {
  if ('date' in item) {
    return item.date;
  }
  // For work/project items, use startDate for sorting
  return item.startDate;
};

export const getRecentContent = (
  limit = 5,
): Array<ContentItem & { type: ContentType }> => {
  return getAllContent()
    .sort((a, b) => getRelevantDate(b).getTime() - getRelevantDate(a).getTime())
    .slice(0, limit);
};

export const getRelatedContent = (
  currentSlug: string,
  currentType: ContentType,
  limit = 3,
): Array<ContentItem & { type: ContentType }> => {
  const allContent = getAllContent();

  // Filter out current item and get items of the same type
  return allContent
    .filter((item) => !(item.slug === currentSlug && item.type === currentType))
    .filter((item) => item.type === currentType)
    .sort((a, b) => getRelevantDate(b).getTime() - getRelevantDate(a).getTime())
    .slice(0, limit);
};

export const getContentByYear = (type?: ContentType) => {
  const content = type ? getCollection(type) : getAllContent();

  return content
    .sort((a, b) => getRelevantDate(b).getTime() - getRelevantDate(a).getTime())
    .reduce(
      (acc, post) => {
        const year = getRelevantDate(post).getFullYear();
        if (!acc[year]) {
          acc[year] = [];
        }
        acc[year].push(post);
        return acc;
      },
      {} as Record<number, typeof content>,
    );
};

export const searchContent = (
  query: string,
  type?: ContentType,
): Array<ContentItem & { type?: ContentType }> => {
  const content = type ? getCollection(type) : getAllContent();
  const searchTerm = query.toLowerCase();

  return content.filter((item) => {
    if ('position' in item) {
      // For work items
      return (
        item.position.toLowerCase().includes(searchTerm) ||
        item.company.toLowerCase().includes(searchTerm) ||
        item.location.toLowerCase().includes(searchTerm)
      );
    } else {
      // For projects and blogs
      return (
        item.title.toLowerCase().includes(searchTerm) ||
        item.description.toLowerCase().includes(searchTerm)
      );
    }
  });
};

export const getContentStats = () => {
  return {
    projects: allProjects.length,
    blogs: allBlogs.length,
    works: allWorks.length,
    total: allProjects.length + allBlogs.length + allWorks.length,
  };
};
