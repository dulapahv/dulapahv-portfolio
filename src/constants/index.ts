export { default as aboutme } from './aboutme.json';
export { default as education } from './education.json';
export { default as experience } from './experience.json';
export { default as project } from './project.json';
export { default as scrollingText } from './scrollingText.json';
