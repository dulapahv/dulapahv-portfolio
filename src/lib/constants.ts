export const BASE_URL =
  process.env.VERCEL_ENV === "development"
    ? "http://localhost:3000/"
    : `https://${process.env.VERCEL_URL}`;
export const ASSETS_URL = `https://assets.dulapahv.dev`;

export const NAME = "Dulapah Vibulsanti";
export const SHORT_NAME = "DulapahV";
export const SITE_NAME = "DulapahV's Portfolio";
export const DESCRIPTION =
  "This website is a personal project to showcase my skills and experience, as well as to share my knowledge and experience with others.";
export const THEME_COLOR = "#fb568a";

export const itemsPerPageOptions = [
  { key: "5", label: "5" },
  { key: "10", label: "10" },
  { key: "15", label: "15" },
];

export const sortByOptions = [
  { key: "start-date-asc", label: "Start date (oldest)" },
  { key: "start-date-desc", label: "Start date (newest)" },
  { key: "end-date-asc", label: "End date (oldest)" },
  { key: "end-date-desc", label: "End date (newest)" },
];

export const typeOptions = [
  { key: "professional", label: "Professional" },
  { key: "personal", label: "Personal" },
];
