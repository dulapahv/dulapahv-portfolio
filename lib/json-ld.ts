import type {
  BlogPosting,
  CreativeWork,
  EmployeeRole,
  Person,
  WebSite,
  WithContext,
} from 'schema-dts';

import { BASE_URL, DESCRIPTION, IS_DEV_ENV, NAME } from '@/lib/constants';
import { social } from '@/lib/social';

import { getAllSkillsForSchema } from './skills-data';

export const personSchema: WithContext<Person> = {
  '@context': 'https://schema.org',
  '@type': 'Person',
  '@id': `${BASE_URL}/#person`,
  name: NAME,
  givenName: 'Dulapah',
  familyName: 'Vibulsanti',
  alternateName: 'DulapahV',
  description: DESCRIPTION,
  gender: 'male',
  nationality: {
    '@type': 'Country',
    name: 'Thailand',
  },
  url: BASE_URL,
  image: {
    '@type': 'ImageObject',
    url: new URL('/avatar.jpg', BASE_URL).toString(),
    caption: 'Dulapah Vibulsanti - Software Engineer',
  },
  sameAs: [...Object.values(social).map(({ href }) => href)],
  alumniOf: [
    {
      '@type': 'CollegeOrUniversity',
      name: 'University of Glasgow',
      url: 'https://www.gla.ac.uk/',
      identifier: 'Russell Group University',
      address: {
        '@type': 'PostalAddress',
        addressLocality: 'Glasgow',
        addressCountry: 'GB',
        postalCode: 'G12 8QQ',
        streetAddress: 'University Avenue',
      },
    },
    {
      '@type': 'CollegeOrUniversity',
      name: "King Mongkut's Institute of Technology Ladkrabang",
      alternateName: 'KMITL',
      url: 'https://www.kmitl.ac.th/',
      address: {
        '@type': 'PostalAddress',
        addressLocality: 'Bangkok',
        addressCountry: 'TH',
        postalCode: '10520',
        streetAddress: '1 Chalong Krung 1 Alley, Lat Krabang',
      },
    },
    {
      '@type': 'EducationalOrganization',
      name: 'Suankularb Wittayalai School',
      url: 'https://www.sk.ac.th/',
      address: {
        '@type': 'PostalAddress',
        addressLocality: 'Bangkok (Wang Burapha Pirom, Phra Nakhon District)',
        addressCountry: 'TH',
        postalCode: '10200',
        streetAddress: '88 Tri Phet Road',
      },
    },
  ],
  jobTitle: 'Software Engineer Graduate',
  worksFor: {
    '@type': 'Organization',
    name: 'NatWest Group',
    url: 'https://www.natwestgroup.com/',
    description:
      "One of the United Kingdom's Big Four banks, serving over 19 million customers",
  },
  workLocation: {
    '@type': 'Place',
    name: 'Edinburgh, Scotland',
    address: {
      '@type': 'PostalAddress',
      addressLocality: 'Edinburgh',
      addressRegion: 'Scotland',
      addressCountry: 'GB',
    },
  },
  knowsAbout: getAllSkillsForSchema(),
  hasCredential: [
    {
      '@type': 'EducationalOccupationalCredential',
      '@id': `${BASE_URL}/#degree-glasgow`,
      name: 'BSc (Hons) Software Engineering - Double Degree',
      description:
        'Double degree program with KMITL. First Class Honours with Specialization in Parallel and Distributed Systems. Final two years completed at University of Glasgow.',
      credentialCategory: 'degree',
      educationalLevel: 'Bachelor',
      dateCreated: '2025-05',
      recognizedBy: {
        '@type': 'CollegeOrUniversity',
        name: 'University of Glasgow',
        url: 'https://www.gla.ac.uk/',
      },
    },
    {
      '@type': 'EducationalOccupationalCredential',
      '@id': `${BASE_URL}/#degree-kmitl`,
      name: 'BEng Software Engineering - Double Degree',
      description:
        'Double degree program with University of Glasgow. First two years completed at KMITL before transferring to Glasgow.',
      credentialCategory: 'degree',
      educationalLevel: 'Bachelor',
      dateCreated: '2025-05',
      recognizedBy: {
        '@type': 'CollegeOrUniversity',
        name: "King Mongkut's Institute of Technology Ladkrabang",
        url: 'https://www.kmitl.ac.th/',
      },
    },
  ],
  knowsLanguage: [
    {
      '@type': 'Language',
      name: 'Thai',
      alternateName: 'th',
    },
    {
      '@type': 'Language',
      name: 'English',
      alternateName: 'en',
    },
  ],
  contactPoint: {
    '@type': 'ContactPoint',
    contactType: 'professional',
    areaServed: 'GB',
    availableLanguage: ['en', 'th'],
    url: `${BASE_URL}/contact`,
  },
  hasOccupation: {
    '@type': 'Occupation',
    name: 'Software Engineer',
    description:
      'A Software Engineer is responsible for designing, developing, testing, and maintaining software applications and systems. They work with various programming languages and technologies to create efficient and scalable solutions.',
    occupationalCategory: '15-1252.00',
    occupationLocation: {
      '@type': 'City',
      name: 'Edinburgh, Scotland',
      address: {
        '@type': 'PostalAddress',
        addressLocality: 'Edinburgh',
        addressRegion: 'Scotland',
        addressCountry: 'GB',
      },
    },
    mainEntityOfPage: {
      '@type': 'WebPage',
      '@id': `${BASE_URL}/#occupation-software-engineer`,
      lastReviewed: '2025-06-29',
    },
  },
};

export const websiteSchema: WithContext<WebSite> = {
  '@context': 'https://schema.org',
  '@type': 'WebSite',
  '@id': `${BASE_URL}/#website`,
  url: BASE_URL,
  name: 'DulapahV Portfolio',
  description: DESCRIPTION,
  publisher: {
    '@id': `${BASE_URL}/#person`,
  },
  inLanguage: 'en-US',
};

export const createBlogPostingSchema = (post: {
  title: string;
  description: string;
  date: Date;
  slug: string;
  readingTime?: string;
  image?: string;
}): WithContext<BlogPosting> => ({
  '@context': 'https://schema.org',
  '@type': 'BlogPosting',
  '@id': `${BASE_URL}/blog/${post.slug}#article`,
  headline: post.title,
  description: post.description,
  datePublished: post.date.toISOString(),
  dateModified: post.date.toISOString(),
  author: {
    '@id': `${BASE_URL}/#person`,
  },
  publisher: {
    '@id': `${BASE_URL}/#person`,
  },
  mainEntityOfPage: {
    '@type': 'WebPage',
    '@id': `${BASE_URL}/blog/${post.slug}`,
  },
  ...(post.image && {
    image: {
      '@type': 'ImageObject',
      url: new URL(post.image, BASE_URL).toString(),
    },
  }),
  ...(post.readingTime && {
    timeRequired: post.readingTime,
  }),
  inLanguage: 'en-US',
  url: `${BASE_URL}/blog/${post.slug}`,
});

export const createProjectSchema = (project: {
  title: string;
  description: string;
  startDate: Date;
  endDate?: Date;
  slug: string;
  image?: string;
}): WithContext<CreativeWork> => ({
  '@context': 'https://schema.org',
  '@type': 'CreativeWork',
  '@id': `${BASE_URL}/project/${project.slug}#project`,
  name: project.title,
  description: project.description,
  creator: {
    '@id': `${BASE_URL}/#person`,
  },
  dateCreated: project.startDate.toISOString(),
  ...(project.endDate && {
    dateModified: project.endDate.toISOString(),
  }),
  ...(project.image && {
    image: {
      '@type': 'ImageObject',
      url: new URL(project.image, BASE_URL).toString(),
    },
  }),
  url: `${BASE_URL}/project/${project.slug}`,
  inLanguage: 'en-US',
});

export const createWorkSchema = (work: {
  position: string;
  company: string;
  location: string;
  startDate: Date;
  endDate?: Date;
  slug: string;
}): WithContext<EmployeeRole> => ({
  '@context': 'https://schema.org',
  '@type': 'EmployeeRole',
  '@id': `${BASE_URL}/work/${work.slug}#role`,
  roleName: work.position,
  startDate: work.startDate.toISOString(),
  ...(work.endDate && {
    endDate: work.endDate.toISOString(),
  }),
});

/* eslint-disable  @typescript-eslint/no-explicit-any */
export const combineSchemas = (schemas: Array<WithContext<any>>): string => {
  return JSON.stringify(
    schemas.length === 1 ? schemas[0] : schemas,
    null,
    IS_DEV_ENV ? 2 : 0,
  );
};
