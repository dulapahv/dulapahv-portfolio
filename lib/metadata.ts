import type { Metadata } from 'next';

import merge from 'deepmerge';

import { BASE_URL } from '@/lib/constants';

type MetadataGenerator = Omit<Metadata, 'description' | 'title'> & {
  title: string;
  description: string;
  ogText?: string;
  ogDescription?: string;
  image?: string;
};

const applicationName = 'DulapahV Portfolio';
const author: Metadata['authors'] = {
  name: 'Dulapah Vibulsanti',
  url: BASE_URL,
};
const publisher = 'Dulapah Vibulsanti';
const twitterHandle = '@dulapahv';

export const createMetadata = ({
  title,
  description,
  ogText,
  ogDescription,
  image,
  ...properties
}: MetadataGenerator): Metadata => {
  const parsedTitle = `${title} | ${applicationName}`;
  const imageUrl =
    image ??
    `/og?title=${encodeURIComponent(ogText ?? title)}&description=${encodeURIComponent(ogDescription ?? description)}`;

  const defaultMetadata: Metadata = {
    title: parsedTitle,
    description,
    applicationName,
    authors: [author],
    creator: author.name,
    formatDetection: {
      telephone: false,
    },
    appleWebApp: {
      capable: true,
      statusBarStyle: 'default',
      title: parsedTitle,
    },
    openGraph: {
      title: parsedTitle,
      description,
      type: 'website',
      siteName: applicationName,
      locale: 'en_US',
      url: imageUrl,
      images: [
        {
          url: imageUrl,
          width: 1200,
          height: 630,
        },
      ],
    },
    publisher,
    twitter: {
      title: parsedTitle,
      description,
      creatorId: twitterHandle,
      card: 'summary_large_image',
      creator: twitterHandle,
      images: [
        {
          url: imageUrl,
          width: 1200,
          height: 630,
        },
      ],
    },
  };

  const metadata: Metadata = merge(defaultMetadata, properties);

  return metadata;
};
