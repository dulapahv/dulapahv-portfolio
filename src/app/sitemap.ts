import { MetadataRoute } from "next";

import { getManyBlog } from "@/data/get-blog";
import { getManyExperience } from "@/data/get-experience";
import { getManyProject } from "@/data/get-project";
import { getManyTag } from "@/data/get-tag";
import { BASE_URL } from "@/lib/constants";
import { ExperienceWithPlace, TagWithStacks } from "@/types/prisma";

export default async function sitemap(): Promise<MetadataRoute.Sitemap> {
  const links = [
    {
      url: BASE_URL,
      lastModified: new Date(),
    },
    {
      url: `${BASE_URL}/experience`,
      lastModified: new Date(),
    },
    {
      url: `${BASE_URL}/project`,
      lastModified: new Date(),
    },
    {
      url: `${BASE_URL}/stack`,
      lastModified: new Date(),
    },
    {
      url: `${BASE_URL}/contact`,
      lastModified: new Date(),
    },
  ];

  const experiences = (
    await getManyExperience({
      select: {
        id: true,
        position: true,
        updatedAt: true,
        place: {
          select: {
            name: true,
          },
        },
      },
    })
  ).item as ExperienceWithPlace[];

  const projects = (
    await getManyProject({
      select: {
        id: true,
        title: true,
        updatedAt: true,
      },
    })
  ).item;

  const blogs = (
    await getManyBlog({
      select: {
        id: true,
        title: true,
        updatedAt: true,
      },
    })
  ).item;

  const tags = (
    await getManyTag({
      select: {
        name: true,
        stacks: {
          select: {
            id: true,
            name: true,
          },
        },
      },
    })
  ).item as TagWithStacks[];

  experiences.forEach((experience) => {
    links.push({
      url: `${BASE_URL}/experience/${experience.id}-${experience.place.name.replace(/ /g, "-")}-${experience.position.replace(/ /g, "-")}`,
      lastModified: experience.updatedAt,
    });
  });

  projects.forEach((project) => {
    links.push({
      url: `${BASE_URL}/project/${project.id}-${project.title.replace(/ /g, "-")}`,
      lastModified: project.updatedAt,
    });
  });

  blogs.forEach((blog) => {
    links.push({
      url: `${BASE_URL}/blog/${blog.id}-${blog.title.replace(/ /g, "-")}`,
      lastModified: blog.updatedAt,
    });
  });

  tags.forEach((stacks) => {
    stacks.stacks.forEach((stack) => {
      links.push({
        url: `${BASE_URL}/stack/${stack.id}-${stack.name.replace(/ /g, "-")}`,
        lastModified: new Date(),
      });
    });
  });

  return links;
}
