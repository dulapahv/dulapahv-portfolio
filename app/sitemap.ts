import type { MetadataRoute } from 'next';

import { BASE_URL } from '@/lib/constants';
import { getCollection, type ContentType } from '@/lib/content-utils';

export default function sitemap(): MetadataRoute.Sitemap {
  const currentDate = new Date();

  // Static pages with high priority
  const staticPages: MetadataRoute.Sitemap = [
    {
      url: BASE_URL,
      lastModified: currentDate,
      changeFrequency: 'monthly',
      priority: 1,
    },
    {
      url: `${BASE_URL}/contact`,
      lastModified: currentDate,
      changeFrequency: 'yearly',
      priority: 0.8,
    },
  ];

  // Type listing pages
  const typePages: MetadataRoute.Sitemap = [
    {
      url: `${BASE_URL}/work`,
      lastModified: currentDate,
      changeFrequency: 'monthly',
      priority: 0.7,
    },
    {
      url: `${BASE_URL}/project`,
      lastModified: currentDate,
      changeFrequency: 'monthly',
      priority: 0.7,
    },
    {
      url: `${BASE_URL}/blog`,
      lastModified: currentDate,
      changeFrequency: 'weekly',
      priority: 0.7,
    },
  ];

  // Generate dynamic content pages
  const types: ContentType[] = ['project', 'blog', 'work'];
  const contentPages: MetadataRoute.Sitemap = types.flatMap((type) =>
    getCollection(type).map((page) => {
      // Determine the URL
      const url = `${BASE_URL}/${type}/${page._meta.path}`;

      // Determine lastModified date based on content type
      let lastModified: Date;
      if ('date' in page && page.date) {
        // Blog posts have a single date
        lastModified = page.date;
      } else if ('startDate' in page && page.startDate) {
        // Work and projects have start/end dates
        if ('endDate' in page && page.endDate) {
          lastModified = page.endDate;
        } else {
          lastModified = page.startDate;
        }
      } else {
        // Fallback to current date
        lastModified = currentDate;
      }

      // Determine change frequency
      let changeFrequency:
        | 'always'
        | 'hourly'
        | 'daily'
        | 'weekly'
        | 'monthly'
        | 'yearly'
        | 'never';
      if (type === 'blog') {
        changeFrequency = 'yearly';
      } else if (type === 'work' || type === 'project') {
        // Check if it's ongoing/present
        const isOngoing = !('endDate' in page && page.endDate);
        changeFrequency = isOngoing ? 'monthly' : 'yearly';
      } else {
        changeFrequency = 'yearly';
      }

      // Determine priority based on type
      const priority = type === 'blog' ? 0.5 : 0.6;

      return {
        url,
        lastModified,
        changeFrequency,
        priority,
      };
    }),
  );

  // Combine all pages
  return [...staticPages, ...typePages, ...contentPages];
}
