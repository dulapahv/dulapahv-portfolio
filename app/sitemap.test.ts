import { describe, expect, it, vi } from 'vitest';

import { BASE_URL } from '@/lib/constants';

import sitemap from './sitemap';

// Mock the content-utils module
vi.mock('@/lib/content-utils', () => ({
  getCollection: vi.fn((type: string) => {
    if (type === 'blog') {
      return [
        {
          _meta: { path: 'test-blog-post' },
          date: new Date('2024-01-15')
        }
      ];
    }
    if (type === 'work') {
      return [
        {
          _meta: { path: 'test-work' },
          startDate: new Date('2023-01-01'),
          endDate: new Date('2024-01-01')
        },
        {
          _meta: { path: 'ongoing-work' },
          startDate: new Date('2024-06-01')
          // No endDate - ongoing
        }
      ];
    }
    if (type === 'project') {
      return [
        {
          _meta: { path: 'test-project' },
          startDate: new Date('2023-05-01'),
          endDate: new Date('2023-08-01')
        }
      ];
    }
    return [];
  })
}));

describe('sitemap', () => {
  it('should return an array of sitemap entries', () => {
    const result = sitemap();

    expect(Array.isArray(result)).toBe(true);
    expect(result.length).toBeGreaterThan(0);
  });

  it('should include static pages', () => {
    const result = sitemap();

    const homePage = result.find(entry => entry.url === BASE_URL);
    const contactPage = result.find(entry => entry.url.includes('/contact'));

    expect(homePage).toBeDefined();
    expect(contactPage).toBeDefined();

    if (homePage) {
      expect(homePage.priority).toBe(1);
      expect(homePage.changeFrequency).toBe('monthly');
    }

    if (contactPage) {
      expect(contactPage.priority).toBe(0.8);
      expect(contactPage.changeFrequency).toBe('yearly');
    }
  });

  it('should include type listing pages', () => {
    const result = sitemap();

    const workPage = result.find(entry => entry.url.endsWith('/work'));
    const projectPage = result.find(entry => entry.url.endsWith('/project'));
    const blogPage = result.find(entry => entry.url.endsWith('/blog'));

    expect(workPage).toBeDefined();
    expect(projectPage).toBeDefined();
    expect(blogPage).toBeDefined();

    [workPage, projectPage, blogPage].forEach(page => {
      if (page) {
        expect(page.priority).toBe(0.7);
      }
    });
  });

  it('should include dynamic content pages', () => {
    const result = sitemap();

    const blogPost = result.find(entry => entry.url.includes('/blog/test-blog-post'));
    const workItem = result.find(entry => entry.url.includes('/work/test-work'));
    const projectItem = result.find(entry => entry.url.includes('/project/test-project'));

    expect(blogPost).toBeDefined();
    expect(workItem).toBeDefined();
    expect(projectItem).toBeDefined();
  });

  it('should set correct priorities for content types', () => {
    const result = sitemap();

    const blogPost = result.find(entry => entry.url.includes('/blog/test-blog-post'));
    const projectItem = result.find(entry => entry.url.includes('/project/test-project'));

    if (blogPost) {
      expect(blogPost.priority).toBe(0.5);
    }

    if (projectItem) {
      expect(projectItem.priority).toBe(0.6);
    }
  });

  it('should set correct change frequency for ongoing work', () => {
    const result = sitemap();

    const ongoingWork = result.find(entry => entry.url.includes('/work/ongoing-work'));
    const completedWork = result.find(entry => entry.url.includes('/work/test-work'));

    if (ongoingWork) {
      expect(ongoingWork.changeFrequency).toBe('monthly');
    }

    if (completedWork) {
      expect(completedWork.changeFrequency).toBe('yearly');
    }
  });

  it('should use correct lastModified dates', () => {
    const result = sitemap();

    const blogPost = result.find(entry => entry.url.includes('/blog/test-blog-post'));
    const workItem = result.find(entry => entry.url.includes('/work/test-work'));

    if (blogPost) {
      expect(blogPost.lastModified).toEqual(new Date('2024-01-15'));
    }

    if (workItem) {
      expect(workItem.lastModified).toEqual(new Date('2024-01-01'));
    }
  });

  it('should have required properties for all entries', () => {
    const result = sitemap();

    result.forEach(entry => {
      expect(entry.url).toBeDefined();
      expect(typeof entry.url).toBe('string');
      expect(entry.lastModified).toBeDefined();
      expect(entry.lastModified).toBeInstanceOf(Date);
      expect(entry.changeFrequency).toBeDefined();
      expect(entry.priority).toBeDefined();
      expect(typeof entry.priority).toBe('number');
    });
  });

  it('should match snapshot', () => {
    const result = sitemap();
    expect(result).toMatchSnapshot();
  });
});
