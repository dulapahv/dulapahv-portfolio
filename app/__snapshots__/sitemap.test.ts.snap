// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sitemap > should match snapshot 1`] = `
[
  {
    "changeFrequency": "monthly",
    "lastModified": 2007-01-09T09:41:00.000Z,
    "priority": 1,
    "url": "https://dulapahv.dev",
  },
  {
    "changeFrequency": "yearly",
    "lastModified": 2007-01-09T09:41:00.000Z,
    "priority": 0.8,
    "url": "https://dulapahv.dev/contact",
  },
  {
    "changeFrequency": "monthly",
    "lastModified": 2007-01-09T09:41:00.000Z,
    "priority": 0.7,
    "url": "https://dulapahv.dev/work",
  },
  {
    "changeFrequency": "monthly",
    "lastModified": 2007-01-09T09:41:00.000Z,
    "priority": 0.7,
    "url": "https://dulapahv.dev/project",
  },
  {
    "changeFrequency": "weekly",
    "lastModified": 2007-01-09T09:41:00.000Z,
    "priority": 0.7,
    "url": "https://dulapahv.dev/blog",
  },
  {
    "changeFrequency": "yearly",
    "lastModified": 2023-08-01T00:00:00.000Z,
    "priority": 0.6,
    "url": "https://dulapahv.dev/project/test-project",
  },
  {
    "changeFrequency": "yearly",
    "lastModified": 2024-01-15T00:00:00.000Z,
    "priority": 0.5,
    "url": "https://dulapahv.dev/blog/test-blog-post",
  },
  {
    "changeFrequency": "yearly",
    "lastModified": 2024-01-01T00:00:00.000Z,
    "priority": 0.6,
    "url": "https://dulapahv.dev/work/test-work",
  },
  {
    "changeFrequency": "monthly",
    "lastModified": 2024-06-01T00:00:00.000Z,
    "priority": 0.6,
    "url": "https://dulapahv.dev/work/ongoing-work",
  },
]
`;
